File: src/main/java/xyz/mizarc/persistentitems/PersistentItems.java
package xyz.mizarc.persistentitems;

import co.aikar.commands.PaperCommandManager;
import org.bukkit.configuration.InvalidConfigurationException;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.plugin.java.JavaPlugin;
import xyz.mizarc.persistentitems.commands.persistentitems.*;
import xyz.mizarc.persistentitems.commands.PersistentItemsCommand;
import xyz.mizarc.persistentitems.listeners.PlayerLoad;
import xyz.mizarc.persistentitems.listeners.PreventItemRemoval;
import xyz.mizarc.persistentitems.listeners.RunLinkedCommand;

import java.io.IOException;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

public class PersistentItems extends JavaPlugin {
    private FileConfiguration config = getConfig();
    private ItemConfigIO itemConfig;
    private ItemContainer itemContainer;
    private PaperCommandManager commandManager;

    @Override
    public void onEnable() {
        super.onEnable();

        transferResources();

        itemConfig = new ItemConfigIO(this);
        itemContainer = new ItemContainer(this);
        itemContainer.loadActiveItems();

        // Listeners
        getServer().getPluginManager().registerEvents(new PlayerLoad(this), this);
        getServer().getPluginManager().registerEvents(new PreventItemRemoval(this), this);
        getServer().getPluginManager().registerEvents(new RunLinkedCommand(this), this);

        commandManager = new PaperCommandManager(this);
        registerLocales();
        registerCommandCompletions();
        registerCommands();
    }

    @Override
    public void onDisable() {
        super.onDisable();
    }

    public ItemConfigIO getItemConfig() {
        return itemConfig;
    }

    public ItemContainer getItemContainer() {
        return itemContainer;
    }

    private void registerCommandCompletions() {
        commandManager.getCommandCompletions().registerCompletion("pitems", context -> {
            Set<Item> items = itemContainer.getAllItems();
            Set<String> itemNames = new HashSet<>();
            for (Item item : items) {
                itemNames.add(item.getId());
            }
            return itemNames;
        });
    }

    private void registerCommands() {
        commandManager.registerDependency(ItemContainer.class, itemContainer);
        commandManager.registerDependency(ItemConfigIO.class, itemConfig);
        commandManager.registerCommand(new PersistentItemsCommand());
        commandManager.registerCommand(new AddCommand());
        commandManager.registerCommand(new RemoveCommand());
        commandManager.registerCommand(new AddCmdCommand());
        commandManager.registerCommand(new ActivateCommand());
        commandManager.registerCommand(new ShowCommand());
        commandManager.registerCommand(new HideCommand());
    }

    private void registerLocales() {
        try {
            commandManager.getLocales().loadYamlLanguageFile("lang/en_US.yml", Locale.ENGLISH);
        } catch (IOException | InvalidConfigurationException error) {
            error.printStackTrace();
        }
    }

    private void transferResources() {
        saveResource("lang/en_US.yml", false);
    }
}

File: src/main/kotlin/xyz/mizarc/persistentitems/PersistentItems.kt
package xyz.mizarc.persistentitems;

import co.aikar.commands.PaperCommandManager;
import org.bukkit.plugin.java.JavaPlugin;
import xyz.mizarc.persistentitems.commands.persistentitems.*;
import xyz.mizarc.persistentitems.commands.PersistentItemsCommand;
import xyz.mizarc.persistentitems.listeners.ItemRemovalListener
import xyz.mizarc.persistentitems.listeners.ItemUseListener;
import xyz.mizarc.persistentitems.listeners.PlayerLoadListener
import java.io.File

import java.util.HashSet;
import java.util.Locale;

class PersistentItems: JavaPlugin() {
    private val config = Config(this)
    private lateinit var commandManager: PaperCommandManager
    private val databaseConnection = DatabaseConnection(this)
    private val itemRepo = ItemRepository("$dataFolder/items")
    private val playerItemsRepo = PlayerItemsRepository(databaseConnection)
    private val persistencyService = PersistencyService(itemRepo)

    override fun onEnable() {
        super.onEnable()

        File("$dataFolder/items").mkdir()
        commandManager = PaperCommandManager(this)
        registerLocales()
        transferResources()
        registerDependencies()
        registerCommandCompletions()
        registerCommands()
        registerListeners()
    }

    override fun onDisable() {
        super.onDisable()
        logger.info("Persistent Items has been disabled.")
    }

    private fun registerLocales() {
        try {
            commandManager.locales.loadYamlLanguageFile("lang/en_US.yml", Locale.ENGLISH)
        }
        catch (error: Exception) {
            error.printStackTrace()
        }
    }

    private fun transferResources() {
        saveResource("lang/en_US.yml", false);
    }

    private fun registerDependencies() {
        commandManager.registerDependency(DatabaseConnection::class.java, databaseConnection)
        commandManager.registerDependency(ItemRepository::class.java, itemRepo)
        commandManager.registerDependency(PlayerItemsRepository::class.java, playerItemsRepo)
        commandManager.registerDependency(PersistencyService::class.java, persistencyService)
    }

    private fun registerCommandCompletions() {
        commandManager.commandCompletions.registerCompletion("pitems") { _ ->
            val items = persistencyService.getActiveItems()
            val itemNames = HashSet<String>()
            for (item in items) {
                itemNames.add(item.name)
            }
            itemNames
        }
    }

    private fun registerCommands() {
        commandManager.registerCommand(PersistentItemsCommand())
        commandManager.registerCommand(AddCommand())
        commandManager.registerCommand(RemoveCommand())
        commandManager.registerCommand(AddCmdCommand())
        commandManager.registerCommand(ActivateCommand())
        commandManager.registerCommand(ShowCommand())
        commandManager.registerCommand(HideCommand())
    }

    private fun registerListeners() {
        server.pluginManager.registerEvents(PlayerLoadListener(playerItemsRepo, persistencyService), this);
        server.pluginManager.registerEvents(ItemRemovalListener(this), this);
        server.pluginManager.registerEvents(ItemUseListener(itemRepo), this);
    }
}
